### ISSUE TITLE
smote

### ISSUE THREAD

**Issue by s01022087242-cloud** [s01022087242-cloud/CN7-#5]
# 🐞 에러 제목

> SMOTE 기반 1:1 리샘플링 데이터(`smote_dataset_1to1.csv`)로 불량 예측 모델 학습 및 평가

---

## 📸 문제 화면 캡쳐

> 실행 화면 대신, 평가 결과 수치를 첨부합니다.

**Random Forest, 80/20 분할 평가 결과**

|              | 예측 정상 | 예측 불량 |
| ------------ | ----- | ----- |
| **실제 정상**    | 1518  | 67    |
| **실제 불량**    | 2     | 1583  |

```text
Accuracy:  0.978
Precision: 0.959
Recall:    0.999
F1:        0.979
```

---

## 🔍 배경 및 접근 방법

* 📍 **대상 데이터:** `labeled_data.csv` 기반 (원본 정상 7,925건 / 불량 71건, 불량 비율 0.9%)
* ⚙️ **적용한 방법:** SMOTE(Synthetic Minority Over-sampling Technique)로 불량 데이터를 71건 → 7,925건까지 오버샘플링하여 정상:불량 = 1:1 비율로 리샘플링
* 🧩 **전처리:** 수치형 변수 정규화(`num__` 접두사) + `EQUIP_NAME` 원-핫 인코딩(`cat__` 접두사) 적용 후 SMOTE 수행
* 🤔 **기대 효과:** 클래스 불균형이 극심한 원본 데이터로는 모델이 불량을 거의 학습하지 못했는데(이전 이슈 참고), 비율을 1:1로 맞추면 정밀도·재현율이 개선될 것으로 예상

---

## 🛠️ 적용 과정

1. 🔧 원본 라벨 데이터(`labeled_data.csv`)에서 수치형 변수 스케일링 + 범주형 변수(`EQUIP_NAME`) 인코딩
2. 🔧 SMOTE로 불량 데이터를 정상 개수(7,925건)와 동일하게 오버샘플링 → `smote_dataset_1to1.csv` 생성
3. 🔧 생성된 데이터를 `train_test_split(test_size=0.2, stratify=y)`으로 분할
4. 🔧 `RandomForestClassifier(n_estimators=300, max_depth=8, min_samples_leaf=2)`로 학습 (class_weight 미적용, 이미 1:1이므로 불필요)

### 💻 사용한 코드

```python
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y)

model = RandomForestClassifier(
    n_estimators=300, max_depth=8, min_samples_leaf=2, random_state=42)
model.fit(X_train, y_train)
```

---

## ✅ 결과 체크리스트

* [x] 🎯 정상:불량 1:1 비율 데이터 생성 완료
* [x] ▶️ 모델 학습 및 평가 정상 실행
* [x] 📈 Precision/Recall/F1 모두 이전 대비 큰 폭으로 개선
* [ ] 🧪 실제 미지의 불량 데이터에 대한 일반화 성능 검증 (미완료, 아래 한계 참고)

---

## ⚠️ 알려진 한계 (반드시 함께 읽어야 함)

이 수치는 액면 그대로 받아들이면 안 됩니다. 아래 문제가 확인되었습니다.

* **데이터 누수(Data Leakage) 발생**: 원본 불량 71건을 7,925건으로 약 **112배** 증식했기 때문에, 진짜 불량 1건 주변에 평균 112개의 매우 유사한(보간된) 합성 샘플이 밀집합니다. train/test를 무작위로 나눠도 테스트셋에 들어간 불량 샘플의 **97.6%가 학습셋의 어떤 샘플과 사실상 동일한 지점**(유클리드 거리 0.05 이내)에 위치함을 확인했습니다.
* **의미**: 모델이 "불량 패턴을 일반화해서 학습"한 게 아니라 "이미 학습에서 본 것과 거의 같은 점을 다시 알아본 것"에 가깝습니다. 따라서 F1 0.979는 **새로운(미지의) 불량을 탐지하는 실제 성능을 대표하지 않습니다.**
* **재현 시 유의점**: SMOTE를 적용한 뒤 그 데이터를 통째로 train/test 분할하는 방식은 이 프로젝트의 평가 방법으로 채택하지 않습니다. 향후 SMOTE를 다시 시도할 경우, 반드시 (1) 원본 데이터를 먼저 train/test로 분리하고 (2) train 쪽에만 SMOTE를 적용해, test는 합성되지 않은 순수 원본 데이터로 유지해야 합니다.

---

## 📝 최종 정리

| 항목       | 내용                                                                 |
| -------- | -------------------------------------------------------------------- |
| 🐞 작업    | `smote_dataset_1to1.csv`(SMOTE로 1:1 리샘플링)로 Random Forest 학습·평가       |
| 📈 결과    | Accuracy 0.978 / Precision 0.959 / Recall 0.999 / F1 0.979              |
| ⚠️ 한계    | 원본 불량 71건을 112배 증식하며 발생한 데이터 누수로, 테스트셋 불량의 97.6%가 학습셋과 근접 중복. 실제 일반화 성능 지표로 사용 불가 |
| 🔁 후속 조치 | SMOTE를 train 분할 이후에만 적용하는 방식으로 재실험 필요 (누수 없는 정직한 평가 확보)               |
